Add BHASA LINDSEA scenarios

raileymontalan opened the pull request:
Add the following for [BHASA](https://arxiv.org/pdf/2309.06085v1):
- [x] Linguistic Diagnostics: syntax minimal pairs, pragmatic reasoning

yifanmai commented on src/helm/benchmark/metrics/bhasa_metrics.py:
```diff
@@ -0,0 +1,316 @@
+
+    def _compute_chrf(self, refs: List[str], pred: str) -> Dict[str, float]:
+        metrics: Dict[str, float] = {}
+        metrics['ChrF++'] = self.chrf_scorer.sentence_score(pred, refs).score
```
Generally our convention for metrics is camel case - could you make this `"chr_f_plus_plus"` instead?

yifanmai replied:
Correction - convention is snake case, not camel case.

yifanmai commented on src/helm/benchmark/metrics/bhasa_metrics.py:
```diff
@@ -0,0 +1,316 @@
+            aggregator.add_scores(self.rouge_scorer.score(ref, pred))
+        aggregates = aggregator.aggregate()
+        for key, value in self.rouge_metrics.items():
+            metrics[value] = aggregates[key].mid.fmeasure * 100
```
Is this to make the range 0 to 100 instead of 0 to 1? We generally prefer the 0 to 1 range.

yifanmai commented on src/helm/benchmark/metrics/bhasa_metrics.py:
```diff
@@ -0,0 +1,316 @@
+            metrics[value] = aggregates[key].mid.fmeasure * 100
+        return metrics
+
+    def _remove_braces(self, text: str) -> str:
```
Optional: Don't know if this is important, but if you want to ensure that braces are removed in a balanced way, then you should do 

```python
        if text.startswith("{") and text.endswith("}"):
            text = text[1:-1]
```
otherwise you might strip the brace from only the start or only the end. Likewise for the other occurrence of this function.

Also, for my education, why do we need to remove braces?

yifanmai commented on src/helm/benchmark/metrics/bhasa_metrics.py:
```diff
@@ -0,0 +1,316 @@
+            text = text[:-1]
+        return text
+    
+    def evaluate(
```
nit: can omit this definition, if all you are doing is calling the super.

yifanmai commented on src/helm/benchmark/metrics/bhasa_metrics.py:
```diff
@@ -0,0 +1,316 @@
+
+        return result
+
+def get_bhasa_machine_translation_metric_specs() -> List[MetricSpec]:
```
Move these to a different file `bhasa_metric_specs.py` (follow the conventions in `common_metric_specs.py`).

The reason for doing this is that we don't want the `bhasa_run_specs` file to import` bhasa_metrics`, which transitively imports optional dependencies, otherwise this would cause `helm-run` to fail for someone who doesn't have the optional dependencies installed.

yifanmai commented on src/helm/benchmark/metrics/bhasa_metrics.py:
```diff
@@ -0,0 +1,316 @@
+        if self.language == "th":
+            return word_tokenize(normalized_text, engine="newmm")
+        elif self.language == "en":
+            return white_space_fix(remove_articles(normalized_text)).split()
```
optional: you can skip `white_space_fix()` by doing `re.split("\s+", remove_articles(normalized_text))`, which would also allow you to delete the definition of `white_space_fix()`.

weiqipedia replied:
Thanks for the suggestion! I was testing this out with the string "  abc  abc  abc  " , and it seems like we get different results when using each method:

`\s+`: [" ", "abc", "abc", "abc", " "]

`white_space_fix()`: ["abc", "abc", "abc"]

But besides this difference, actually this code was largely taken from the SQuAD v1.1 evaluation code, and I guess we were trying to preserve as much of the original as possible for transparency. What do you think?

yifanmai replied:
Sounds good.

Optionally: it would be nice to add a comment to the code that this is to match the SQuAD v1.1 behavior.

yifanmai commented on src/helm/benchmark/metrics/bhasa_metrics.py:
```diff
@@ -0,0 +1,316 @@
+        def remove_articles(text: str) -> str:
+            return re.sub(r"\b(a|an|the)\b", " ", text)
+
+        def white_space_fix(text: str) -> str:
```
optional: can maybe delete this - see later comment.

yifanmai commented on src/helm/benchmark/metrics/bhasa_metrics.py:
```diff
@@ -0,0 +1,316 @@
+
+            for name, metric in self.metrics.items():
+                name = MetricName(name)
+                metric = cast(Callable[[str, str], float], metric)
```
shouldn't need this cast - put the type in the declaration of `self.metrics`

yifanmai commented on src/helm/benchmark/metrics/bhasa_metrics.py:
```diff
@@ -0,0 +1,316 @@
+
+    def __init__(self, language: str = 'en'):
+        self.language: str = language
+        self.metrics: Dict[str, Callable] = {
```
I think the type should be `Dict[str, Callable[[str, str], float]]`

yifanmai commented on src/helm/benchmark/metrics/bhasa_metrics.py:
```diff
@@ -0,0 +1,316 @@
+                name = MetricName(name)
+                metric = cast(Callable[[str, str], float], metric)
+                score_1 = max(metric(gold.output.text.strip(), preds[0]) for gold in golds)
+                score_k = max(metric(gold.output.text.strip(), pred) for gold in golds for pred in preds)
```
nit: can move this immediately after `if adapter_spec.num_outputs != 1:` to avoid making an unnecessary extra call to `metric()` when `preds` is length 1.

yifanmai commented on src/helm/benchmark/metrics/bhasa_metrics.py:
```diff
@@ -0,0 +1,316 @@
+    def __init__(self, language: str = 'en'):
+        self.language: str = language
+        self.rouge_metrics = {
+            "rougeL": "xlsum_rouge_l",
```
nit: prefer "rouge_l" to maintain conventions.

yifanmai commented on src/helm/benchmark/scenarios/bhasa_scenario.py:
```diff
@@ -0,0 +1,1908 @@
+                data = df[df["passage_text"].apply(len) < df["passage_text"].apply(len).quantile(.2)]
+            else:
+                # Sample 100 examples for test
+                data = df.sample(n=100, random_state=5018)
```
100 seems a bit small, could we do 1k or 10k? Or does this cause memory issues.

weiqipedia replied:
You're right, we initially used a small number for QA and summarization because the texts were very long and we had limited resources, but ideally a larger number would be better for more statistical power.

I initially thought we might want to keep it as close to what we used in the BHASA paper as possible, but I suppose we could just apply the improvement now too.

The maximum we can do here would be 565 (the size of the validation set).

yifanmai replied:
OK - also note that `helm-run` also does downsampling if you pass in a flag like `--max-eval-instances 100`.

raileymontalan replied:
Hi @weiqipedia, to clarify, should we increase the sample size to the maximum of the test set for TyDiQA only? Or also do the same for the other scenarios (such as XQuAD and IndicQA?)

weiqipedia replied:
We can increase the sampling to 1000 for XQuAD and IndicQA, and no need to sample for TyDiQA since we are just taking the entire validation set

yifanmai commented on src/helm/benchmark/scenarios/bhasa_scenario.py:
```diff
@@ -0,0 +1,1908 @@
+        df = dataset.to_pandas()
+
+        # Sample 100 examples for test
+        df_test = df.sample(n=100, random_state=self.map[self.language]["random_state"])
```
likewise

weiqipedia replied:
We can increase this to 1000 and still have 190 remaining for 5-shot examples, although we might then want to consider putting the ones with the shortest lengths in the 190.

yifanmai commented on src/helm/benchmark/scenarios/bhasa_scenario.py:
```diff
@@ -0,0 +1,1908 @@
+
+            if split == "train":
+                # Select only bottom 20th percentile by length for in-context examples as examples are very long
+                data = df[df["passage_text"].apply(len) < df["passage_text"].apply(len).quantile(.2)]
```
Is there possible overlap between the train and test splits? I noticed `XQuADScenario` has a fix that avoids this - is this also needed here?

weiqipedia replied:
TyDiQA does have a train split and validation split, so we use the train for train and validation for test.

yifanmai commented on src/helm/benchmark/run_specs/bhasa_run_specs.py:
```diff
@@ -0,0 +1,678 @@
+    },
+}
+
+def generate_xquad_run_spec(language="th"):
```
optional: can inline this into `get_xquad_spec()` since it isn't used elsewhere

yifanmai commented on src/helm/benchmark/run_specs/bhasa_run_specs.py:
```diff
@@ -0,0 +1,678 @@
+        name=name,
+        scenario_spec=scenario_spec,
+        adapter_spec=adapter_spec,
+        metric_specs=get_f1_metric_specs(),
```
For all the sentiment analysis scenarios, you probably want `get_exact_match_metric_specs() + get_classification_metric_specs()` - the confusingly-named `get_f1_metric_specs()` gives you word overlap F1 instead of classification F1 and is intended for open ended generation evaluation.

yifanmai commented on src/helm/benchmark/run_specs/bhasa_run_specs.py:
```diff
@@ -0,0 +1,678 @@
+        name=name,
+        scenario_spec=scenario_spec,
+        adapter_spec=adapter_spec,
+        metric_specs=get_f1_metric_specs(),
```
Same comment regarding `metric_specs` as above.

yifanmai commented on src/helm/benchmark/run_specs/bhasa_run_specs.py:
```diff
@@ -0,0 +1,678 @@
+    }
+}
+
+def generate_xlsum_run_spec(language="id"):
```
optional: inline

same for the other inline-able functions below

yifanmai commented on src/helm/benchmark/metrics/xlsum/README.md:
Could you clarify:

- Is this subfolder is a redistributed copy of another code repository? If so:
    - What is the URL for the original source?
    - Has it been modified from the original source?

I see some URLs in `README.md` and source code, but it is hard for me to tell if these URLs refer to the original source, or to other redistributed code contained in the original source.

yifanmai reviewed (changes requested):
This looks amazing! I mostly just have minor comments. It is a little long so I did a partial review (not quite done with the scenarios, schema and run entries yet) - will review the rest tomorrow.

raileymontalan commented:
Hi @yifanmai, apologies for the mixup. We've decided to keep only the LINDSEA-related code here. All other code has been moved to [this PR](https://github.com/stanford-crfm/helm/pull/2648). I will be compiling and addressing your comments there.

yifanmai commented on src/helm/benchmark/run_specs/bhasa_run_specs.py:
```diff
@@ -0,0 +1,108 @@
+@run_spec_function("lindsea_syntax_minimal_pairs")
+def get_lindsea_syntax_minimal_pairs_spec(language: str = "id", method: str = "mcq") -> RunSpec:
+    name = f"lindsea_syntax_minimal_pairs_{language}"
+    if method == "mcq":
```
What's the reason for doing the "A" and "B" formatting yourself, instead of using `get_multiple_choice_joint_adapter_spec()` which will do it for you? `get_multiple_choice_joint_adapter_spec()` has the added advantage of being more similar to `get_multiple_choice_separate_adapter_spec()` except for logprobs vs generation.

yifanmai commented on src/helm/benchmark/static/schema_bhasa.yaml:
```diff
@@ -0,0 +1,167 @@
+      - efficiency
+      - general_information
+    environment:
+      main_name: accuracy
```
This should be `exact_match` or `quasi_exact_match`.

yifanmai commented on src/helm/benchmark/static/schema_bhasa.yaml:
```diff
@@ -0,0 +1,167 @@
+---
+############################################################
+metrics:
```
Add `exact_match` or `quasi_exact_match` to this list.

yifanmai reviewed (approved):
Mostly looks good, just some minor stuff left.

yifanmai commented:
Please run the linter: at the root of the repo, run:

```sh
pip install -e '.[dev]'
./pre-commit.sh
```

yifanmai reviewed (commented):
Could you merge / rebase the changes from #2648?

raileymontalan commented:
Hi @yifanmai, I saw that a lot of PR checks (including this one) were failing with the same error message (Failed to build backports.zoneinfo). Please advise.

yifanmai commented:
@raileymontalan sorry, the main branch was broken. Could you merge main again? That should pick up the fix.

raileymontalan commented:
@yifanmai Thanks! This PR has now passed all the checks :)

weiqipedia commented:
As a reminder, there is still a need to discuss and revamp the calculation of metrics for the LINDSEA scenarios before this can be merged!

yifanmai commented:
I'll merge this first so that I can do some prototyping on top of this.
